Allow to transfer all Thread datasets with TLV

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
This commit allows to transfer all Thread datasets with TLV. Since PR #22022 the preferred dataset is transmitted when using Matter external commissioning. This commit makes the Thread configuration dialog to have feature parity.

It also allows to select the preferred Thread border router. This can help to add a new dataset with an existing border router. Otherwise the Apple Matter flow might complain with the following error:

![add-matter-device-non-existing-border-agent-id](https://github.com/user-attachments/assets/fb4a278e-635c-47a7-a94c-e7d047f9a681)



## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

agners committed:
Allow to transfer all Thread datasets with TLV

This commit allows to transfer all Thread datasets with TLV. Since
PR #22022 the preferred dataset is transmitted when using Matter
external commissioning. This commit makes the Thread configuration
dialog to have feature parity.

bramkragten commented on src/panels/config/integrations/integration-panels/thread/thread-config-panel.ts:
```diff
@@ -208,8 +208,15 @@ export class ThreadConfigPanel extends SubscribeMixin(LitElement) {
-              const showOverflow =
-                ("dataset" in network && router.border_agent_id) || otbr;
+              const showDefaultRouter = !!network.dataset;
+              const isDefaultRouter =
```
Should also be used to display the icon in the row on line [245](https://github.com/home-assistant/frontend/pull/22183/files#diff-796b8571e876e5bf8f287e5037afd3219ca57fdb374673ff800e09d63104b087R245)

bramkragten commented on src/panels/config/integrations/integration-panels/thread/thread-config-panel.ts:
```diff
@@ -208,8 +208,15 @@ export class ThreadConfigPanel extends SubscribeMixin(LitElement) {
+              const showDefaultRouter = !!network.dataset;
+              const isDefaultRouter =
+                showDefaultRouter &&
+                ((network.dataset!.preferred_border_agent_id &&
+                  router.border_agent_id ===
+                    network.dataset!.preferred_border_agent_id) ||
```
Why did we add border agent id to this?

agners replied:
Hm, yeah not sure. I think I had a probably old datastore which had the preferred border agent id set but not the preferred extended address :thinking: . But today we have to rely more on the extended address since Apple still does not have a border agent id :cry: .

agners replied:
I'll drop the border_agent_id check.

bramkragten commented on src/panels/config/integrations/integration-panels/thread/thread-config-panel.ts:
```diff
@@ -467,10 +487,7 @@ export class ThreadConfigPanel extends SubscribeMixin(LitElement) {
-      network.dataset && router.border_agent_id
-        ? Number(ev.detail.index)
-        : Number(ev.detail.index) + 1;
+    const index = Number(ev.detail.index);
```
So we are sure we always have a dataset for OTBR networks right?

agners replied:
Pretty much, yes: The config flow of the OTBR integration makes sure that the OTBR has an active dataset: If there is a preferred dataset with a TLV in our store, it joins that network, otherwise it creates a new one. The active dataset is then added to the Thread store on OTBR config entry setup.

agners committed:
Drop preferred border agent id as additional metric for default router

We always have the extended address, so use this as primary and only
metric which router is the default. The preferred border agent id gets
updated best effort.

Also use isDefaultRouter consistently in the code.

darkxst commented on src/panels/config/integrations/integration-panels/thread/thread-config-panel.ts:
```diff
@@ -467,10 +482,7 @@ export class ThreadConfigPanel extends SubscribeMixin(LitElement) {
-      network.dataset && router.border_agent_id
-        ? Number(ev.detail.index)
-        : Number(ev.detail.index) + 1;
+    const index = Number(ev.detail.index);
```
@agners I think this change is the cause of Issue #22306 

`network.dataset` is not set there and the indexing on the resulting menu is off by 1

agners replied:
Hm, we discussed this above, and from what I can tell every OTBR entry should actually have a dataset :thinking: 

https://github.com/home-assistant/frontend/pull/22183#discussion_r1784383134

darkxst replied:
It would seem in this case that the router has not been assigned to any network yet. I suppose that is why its missing a dataset? 

[config_entry-thread-4f46c2ed6336e72b15a52386aa9ca930 (3).json](https://github.com/user-attachments/files/17327190/config_entry-thread-4f46c2ed6336e72b15a52386aa9ca930.3.json)

agners replied:
Yeah so that is the thing: From what I understand how the OTBR integration works, it will either join the preferred network or create a new one. Looking at the code I don't understand how you can get into any other state :see_no_evil:  I must misunderstand something :thinking: 

Did you add that OTBR recently?

darkxst replied:
I added a new device to the odroid.local OTBR today (well previously it should have been configured with a different device but been disconnected for a while and preferred network was already set as well).  Bringing up the new device did not add it to preferred network.

agners replied:
I just can't reproduce this :see_no_evil:  For me, the OTBR integration provides a dataset.

So I've stopped an OTBR add-on running on a different system, deleted the integration and credentials on my test system. I then added the OTBR again. But this immediately imported the credentials.

The only way I can imagine that to fail is if somehow fetching the credentials or setting up a new Thread network fails. But this should lead to log entries in your Core when adding the OTBR integration :thinking:  Are the Core logs clean?

Are you using OTBR in a container? If so which OTBR POSIX hash is that built from? The behavior really should be the same since all patches are upstream.

In the end, I can try to fix the frontend by "feel", but I really would like to understand how things get there, and actually test my fix :see_no_evil:

agners replied:
Hm, ok I see, we setup a new active dataset only when the config entry gets created. So if the OTBR integration is there, but the add-on/container cleared it's network config, then we just try to read the dataset, which won't return a valid dataset.

darkxst replied:
In this case was using HA addon. The integration was previously configured and active, but addon had been offline for atleast a few weeks or more (I loose track as too many instances and test setups running).  Normally it keeps the dataset when swapping to new radio device so not sure what happened to the dataset in first place!

 I just deleted and reinstalled integration now and it added device to preferred network automatically. I have however in the past had to manually use the "add to preferred" .
